### PR TITLE
i.MX6 rework which decouples GUI (fb0) and video rendering (fb1), framebuffers are composed with DP in hw

### PR DESCRIPTION
**Pull request by smallint**
This is another rework of the i.IMX6 decoding and rendering path that supersedes #5805. More details in http://forum.kodi.tv/showthread.php?tid=211289&pid=1892440#pid1892440.

Excerpt of architecture:
- Instead of render VPU decoded buffers with GPU (which involves expensive memory copy operations in combination with de-interlacing) we now render exclusively to another hardware framebuffer (fb1)
- Rendering directly to this new framebuffer can be combined with de-interlacing in one step
- DP combines both framebuffers in hardware with alpha blending (if 32bpp) or colour keying (if 16bpp)
- Together with the optimizations brought in by @FernetMenta regarding video rendering we can now even achieve double rate de-interlacing of 1080i streams
- In a nutshell: video rendering with GPU is bad, IPU is good

What does not work yet:
- ~~RenderCapture~~

TODOs:
- Find a better place for class CIMXContext which is currently part of the decoder implementation but required on some other places as well

Recommendations for client setup:
- enable VPU@352Mhz in kernel
- setup 32 bit framebuffer in uEnv.txt (bpp=32)
- disable console blanking in uEnv.txt (consoleblank=0)

Have fun!

### CODE REVIEW

**Comment by smallint** [comment]
One additional note: algorithmdirtyregions must be set to either 0 or 3. 1 and 2 do not work for the time being. I suppose that is a side effect of the hardware clipping work in progress.

File: xbmc/cores/VideoRenderers/LinuxRendererGLES.cpp [hunk 1]
```diff
@@ -618,7 +663,6 @@ void CLinuxRendererGLES::FlipPage(int source)
     m_iYV12RenderBuffer = NextYV12Texture();
 
   m_buffers[m_iYV12RenderBuffer].flipindex = ++m_flipindex;
-
```
Comment (FernetMenta) [inline xbmc/cores/VideoRenderers/LinuxRendererGLES.cpp]: what editor do you use? please drop those unwanted changes.

File: xbmc/cores/VideoRenderers/LinuxRendererGLES.cpp [hunk 2]
```diff
@@ -1253,7 +1291,7 @@ void CLinuxRendererGLES::RenderMultiPass(int index, int field)
 //    imgwidth  *= planes[0].pixpertex_x;
 //    imgheight *= planes[0].pixpertex_y;
 //  }
-//  
+//
```
Comment (FernetMenta) [inline xbmc/cores/VideoRenderers/LinuxRendererGLES.cpp]: same here

**Comment by FernetMenta** [comment]
this is mostly code exclusively used by MX6 which you know best. the other code looks ok to me.

File: xbmc/cores/VideoRenderers/LinuxRendererGLES.cpp (hunk 1 above)
Comment (smallint) [inline xbmc/cores/VideoRenderers/LinuxRendererGLES.cpp]: qtcreator, it stripes unnecessary spaces when saving (but does not remove lines). If that is a complete line that has been removed then it was my work. I will add it again ;)

**Comment by fritsch** [comment]
On the raspberry PI the gui update is slowed down, when in video mode. I tried this afternoon to apply the specific patch also to the imx work, but it works a bit different.

Is there a chance that we e.g. throttle the OSD rendering, codec screen, when playing fullscreen video?

That would help imx, wetek and all those a lot. But is slightly off topic here.

**Comment by smallint** [comment]
@koying, it would be interesting if that PR works on Android as well. If not it is most likely just a matter of device names which are different there. If it won't work at all I could still activate the old path and your Bob deinterlacer for Android. I would like to avoid to break Android support with that PR.

**Comment by wolfgar** [comment]
Thanks a lot smallint for this PR.
I am definitively convinced that this approach is the one which gets the best out of the imx6 architecture...

**Comment by koying** [comment]
@smallint Would you mind adding a high level description of the new architecture in the PR header, please.
Both for my comprehension and for future reference ;)

**Comment by smallint** [comment]
@koying, I tried ... OK?

**Comment by koying** [comment]
Perfect thanks. Re your initial question, it works the same way with amlogic, so if the permissions are set right in the FW, we should be able to do the same thing on android.

**Comment by smallint** [comment]
Latest changes and impressive improvements from @wolfgar merged (thanks) and tested. RenderCapture was added and this PR is now complete.

**Comment by fritsch** [comment]
Then let's build it. Thank you very much for bringing IMX to a new level.

jenkins build this please

**Comment by wolfgar** [comment]
Thanks a lot for your tremendous work on this rendering refactoring @smallint 
I have just tested that the current rebase works fine
To me it is all good for merge...

**Comment by fritsch** [comment]
Oki - final build after laster build and that's wait for @FernetMenta final ack.

jenkins build this please

**Comment by wolfgar** [comment]
To fix little remaining build issue : https://github.com/wolfgar/xbmc/commit/e9b6dfeb62ed7f97516a2eb33327f409de8f029f

**Comment by fritsch** [comment]
Little compiler warning: https://github.com/smallint/xbmc/blob/1597ceda6e8c9ca8355e970d8e261d61ee7afcad/xbmc/cores/dvdplayer/DVDCodecs/Video/DVDVideoCodecIMX.cpp#L1483

this should be void or return some bool?

**Comment by FernetMenta** [comment]
+1

**Comment by fritsch** [comment]
jenkins build this please

**Comment by MartijnKaijser** [comment]
feature merge window is closed so this can go in 1 March

**Comment by fritsch** [comment]
Sorry for yet another build, did not see that: https://github.com/wolfgar/xbmc/commit/e9b6dfeb62ed7f97516a2eb33327f409de8f029f was not yet picked.

**Comment by FernetMenta** [comment]
we agreed that we start Match 1st with new process. currently we don't have a merge window, means this can go in when ready. I also want the pvr binary addon PR merged soon because it blocks pvr development.

**Comment by smallint** [comment]
I picked the fix from @wolfgar and removed the compiler warning. Hope it builds now.

**Comment by fritsch** [comment]
jenkins build this please

**Comment by wolfgar** [comment]
seems the build issue is unrelated to this commit so we may be fine ?

**Comment by MartijnKaijser** [comment]
jenkins build this please

**Comment by fritsch** [comment]
Thanks to everyone involved!
